modules/backup-azure/client.go

```diff
@@ -0,0 +1,192 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2023 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package modstgazure
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/bloberror"
+	"github.com/pkg/errors"
+	"github.com/weaviate/weaviate/entities/backup"
+)
+
+type azureClient struct {
+	client     *azblob.Client
+	config     clientConfig
+	serviceURL string
+	dataPath   string
+}
+
+func newClient(ctx context.Context, config *clientConfig, dataPath string) (*azureClient, error) {
+	connectionString := os.Getenv("AZURE_STORAGE_CONNECTION_STRING")
+	if connectionString != "" {
+		client, err := azblob.NewClientFromConnectionString(connectionString, nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "create client using connection string")
+		}
+		return &azureClient{client, *config, "", dataPath}, nil
+	}
+
+	// Your account name and key can be obtained from the Azure Portal.
+	accountName := os.Getenv("AZURE_STORAGE_ACCOUNT_NAME")
+	accountKey := os.Getenv("AZURE_STORAGE_PRIMARY_ACCOUNT_KEY")
+
+	if accountName == "" {
+		return nil, errors.New("AZURE_STORAGE_ACCOUNT_NAME must be set")
+	}
+
+	// The service URL for blob endpoints is usually in the form: http(s)://<account>.blob.core.windows.net/
+	serviceURL := fmt.Sprintf("https://%s.blob.core.windows.net/", accountName)
+
+	if accountKey != "" {
+		cred, err := azblob.NewSharedKeyCredential(accountName, accountKey)
+		if err != nil {
+			return nil, err
+		}
+
+		client, err := azblob.NewClientWithSharedKeyCredential(serviceURL, cred, nil)
+		if err != nil {
+			return nil, err
+		}
+		return &azureClient{client, *config, serviceURL, dataPath}, nil
+	}
+
+	client, err := azblob.NewClientWithNoCredential(serviceURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &azureClient{client, *config, serviceURL, dataPath}, nil
+}
+
+func (a *azureClient) HomeDir(backupID string) string {
+	return a.serviceURL + path.Join(a.config.Container, a.makeObjectName(backupID))
+}
+
+func (a *azureClient) makeObjectName(parts ...string) string {
+	base := path.Join(parts...)
+	return path.Join(a.config.BackupPath, base)
+}
+
+func (a *azureClient) GetObject(ctx context.Context, backupID, key string) ([]byte, error) {
+	objectName := a.makeObjectName(backupID, key)
+
+	if err := ctx.Err(); err != nil {
```

Comment:
I don't think this check is needed.
The context is very unlikely to be canceled and if it was canceled `a.client.DownloadStream` would take care of it.